active/0014-rolling-cluster-upgrade.md

```diff
@@ -0,0 +1,122 @@
+# Rolling cluster upgrade
+
+## Change log
+
+* 2021-09-21: @k32 Initial draft
+* 2021-09-23: @k32 Applied remarks
+
+## Abstract
+
+Currently EMQ X upgrade procedure has two modes of upgrade:
+
+1. Patch releases.
+   Upgrading between the patch versions is done by patching the live nodes using Erlang hot code patching feature.
+   The patch is loaded to the existing running nodes.
+   This is a low-maintenance upgrade path, however severely limited in the scope of the changes.
+1. Minor and major releases requires taking down the entire cluster and redirecting the MQTT traffic to a different cluster running the new version of the software.
+   This approach introduces a lot of operational overhead.
+
+EMQ X should support rolling upgrade of the cluster when nodes (or pods) are taken down and replaced one by one for minor version upgrades.
+Patch version upgrades will remain the same.
+
+This EIP introduces the guidelines for writing the backward- and forward-compatible code.
+Also it documents the necessary changes to the existing code base.
+
+## Motivation
+
+In order to make upgrading the cluster smoother, EMQ X should support rolling cluster upgrades and green-blue deployments.
+It should be possible to upgrade the cluster without taking it down entirely.
```

Comment:
Nodes in the cluster could run with different version, right?

Reply:
Yes

Reply:
Maybe we could add it in motivation that we want the nodes in cluster could run with different versions (interchangeable).

Reply:
As long as the cluster is able to roll back to old version if something bad happens in the middle of the rolling upgrade, I think it is good.